abci/src/runtime.rs

```diff
@@ -0,0 +1,101 @@
+//! Abstractions for facilitating runtime-independent code.
+
+#[cfg(feature = "runtime-async-std")]
+pub mod async_std;
+#[cfg(feature = "runtime-tokio")]
+pub mod tokio;
+
+pub use crate::runtime::interface::{
+    ChannelNotify, ClientCodec, Receiver, Sender, ServerCodec, TaskSpawner, TcpListener, TcpStream,
+};
+
+/// Implemented by each runtime we support.
+pub trait Runtime: 'static {
```

Comment:
Interesting, why we don't have a crate for this yet🤔️

Reply:
Technically this `Runtime` trait's somewhat crate-specific right now. But when GATs land in Rust stable I should be able to make it more generic and maybe extract a crate from it. Later today I'll hopefully be working on the non-`async` interface for the `Runtime` trait, which will effectively make this crate usable with just `std` as well.

Do you perhaps know of any other libraries that offer this kind of functionality? I've tried looking, but can't find any, but maybe I'm just not searching for the right terms.

Reply:
There's [async_executors](https://docs.rs/async_executors/) which has similar purpose but it don't covers io apis.
For network stuff, what do you think of make a generic version of [Async<T>](https://github.com/smol-rs/async-io/blob/master/src/lib.rs#L489), and build other abstractions([TcpStream](https://github.com/smol-rs/async-net/blob/master/src/tcp.rs#L319)).

Reply:
Yeah it's hard to find something that'll have the same effect as what I'm looking for right now with this `Runtime` trait.

I think it's actually due to shortcomings in the Rust language in being able to express certain types of generic structures, but again, I'm hoping those shortcomings will be addressed once GATs are available on stable.

I built a little [proof-of-concept using GATs for creating MPSC channel abstractions with Rust nightly](https://github.com/thanethomson/async-channel-abs/blob/master/src/runtime.rs#L11) and it seems to work. This kind of language feature should easily allow for the creation of these kinds of common abstractions over different runtimes.